assembly/src/assembler/mod.rs

```diff
@@ -148,15 +148,33 @@ impl Assembler {
         Ok(Program::with_kernel(program_root, self.kernel.clone(), cb_table))
     }
 
+    /// Compiles the provided [ProgramAst] into a [CodeBlock].
+    pub fn compile_program_ast(&self, program: ProgramAst) -> Result<CodeBlock, AssemblyError> {
+        let (local_procs, body) = program.into_parts();
+
+        let mut context = AssemblyContext::new(AssemblyContextType::Program);
+        // compile all local procedures; this will add the procedures to the specified context
+        for proc_ast in local_procs.iter() {
+            if proc_ast.is_export {
+                return Err(AssemblyError::exported_proc_in_program(&proc_ast.name));
+            }
+            self.compile_procedure(proc_ast, &mut context)?;
+        }
+
+        // compile the program body
+        let program_root = self.compile_body(body.iter(), &mut context, None)?;
+
+        Ok(program_root)
+    }
```

Comment:
The intent of this is to be used for compiling note scripts, right? I'm thinking maybe we should change it back to a version of `compile_in_context` we had before. It doesn't need to be very different from what you have now - e.g., it could look something like this:
```Rust
pub fn compile_in_context(
    &self,
    program: ProgramAst,
    context: &mut AssemblyContext,
) -> Result<CodeBlock, AssemblyError> {
    // TODO: make sure the context is for compiling programs

    let (local_procs, body) = program.into_parts();

    // compile all local procedures; this will add the procedures to the specified context
    for proc_ast in local_procs.iter() {
        if proc_ast.is_export {
            return Err(AssemblyError::exported_proc_in_program(&proc_ast.name));
        }
        self.compile_procedure(proc_ast, context)?;
    }

    // compile the program body
    let program_root = self.compile_body(body.iter(), context, None)?;

    Ok(program_root)
}
```

Then, we can change the `compile()` method to look something like this:
```Rust
  pub fn compile<S>(&self, source: S) -> Result<Program, AssemblyError>
  where
      S: AsRef<str>,
  {
      // parse the program into an AST
      let source = source.as_ref();
      let program = ProgramAst::parse(source)?;

      // compile the program
      let mut context = AssemblyContext::new(AssemblyContextType::Program);
      let program_root = self.compile_in_context(program, &mut context)?;

      // convert the context into a call block table for the program
      let cb_table = context.into_cb_table(&self.proc_cache.borrow());

      // build and return the program
      Ok(Program::with_kernel(program_root, self.kernel.clone(), cb_table))
  }
```

Reply:
I like the idea of calling `compile_in_context(..)` from `compile(..)` but I'm not sure if the caller should be responsible for managing the context.  Maybe instead the `compile_program_ast(..)` should return both `program_root` and `cb_table`.  This would become:

```rust
  pub fn compile<S>(&self, source: S) -> Result<Program, AssemblyError>
  where
      S: AsRef<str>,
  {
      // parse the program into an AST
      let source = source.as_ref();
      let program = ProgramAst::parse(source)?;

      // compile the program
      let mut context = AssemblyContext::new(AssemblyContextType::Program);
      let (program_root, cb_table) = self.compile_program_ast(program, &mut context)?;

      // build and return the program
      Ok(Program::with_kernel(program_root, self.kernel.clone(), cb_table))
  }
```

Are there any benefits to having the caller manage the context?

Reply:
I'm thinking of how we'd use it down the road to construct transaction programs. Whether we do it by manually combining `CodeBlock`'s or via `pcall` instruction in the future, we'd still need to build up a callset of all invoked procedures. `AssemblyContext` already keeps track of this - so, I think it would a natural way to build a unified callset across all notes - though, we'd have to make a few updates before we can use it like this.

Reply:
This makes sense.  I think we will need to make some changes as the `into_cb_table(..)` currently consumes the context but this shouldn't be a problem.